// vlc.js/lib/src/http/classes/VLCStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @class VLCStatus
 * @description Current status about VLC.
 */
var VLCStatus = /** @class */ (function () {
    /**
     * @constructor
     * @param {VLCRequest} vlcRequest
     */
    function VLCStatus(vlcRequest) {
        var parsed = JSON.parse(vlcRequest.data.toString());
        this.fullscreen = (parsed.fullscreen === 0 ? false : parsed.fullscreen);
        this.stats = parsed.stats;
        this.audiodelay = parsed.audiodelay;
        this.apiversion = parsed.apiversion;
        this.currentplid = parsed.currentplid;
        this.time = parsed.time;
        this.volume = parsed.volume;
        this.length = parsed.length;
        this.random = parsed.random;
        this.audiofilters = parsed.audiofilters;
        this.rate = parsed.rate;
        this.videoeffects = parsed.videoeffects;
        this.state = parsed.state;
        this.loop = parsed.loop;
        this.version = parsed.version;
        this.position = parsed.position;
        this.information = parsed.information;
        this.repeat = parsed.repeat;
        this.subtitledelay = parsed.subtitledelay;
        this.equalizer = parsed.equalizer;
        this.date = parsed.date;
    }
    return VLCStatus;
}());
exports.VLCStatus = VLCStatus;
var VLCPlaylistStatus;
(function (VLCPlaylistStatus) {
    VLCPlaylistStatus["stopped"] = "stopped";
    VLCPlaylistStatus["playing"] = "playing";
    VLCPlaylistStatus["paused"] = "paused";
    VLCPlaylistStatus["unknown"] = "unknown";
})(VLCPlaylistStatus = exports.VLCPlaylistStatus || (exports.VLCPlaylistStatus = {}));
